Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1162 @@
+    public class TypeInfoWrapper : ComTypes.ITypeInfo, IDisposable
+    {
+        private DisposableList<TypeInfoWrapper> _typeInfosWrapped;
+        private TypeLibWrapper _containerTypeLib;
+        public TypeLibWrapper Container { get => _containerTypeLib; }
+        private int _containerTypeLibIndex;
+        private bool _isUserFormBaseClass = false;
+        private IntPtr _rawObjectPtr;
+        private ComTypes.ITypeInfo _wrappedObjectRCW;
+
+        private ComTypes.TYPEATTR _cachedAttributes;
+        public ComTypes.TYPEATTR Attributes { get => _cachedAttributes; }
+
+        private RestrictComInterfaceByAggregation<ComTypes.ITypeInfo> _ITypeInfo_Aggregator;
+        private ComTypes.ITypeInfo target_ITypeInfo { get => _ITypeInfo_Aggregator?.WrappedObject ?? _wrappedObjectRCW; }
+
+        private RestrictComInterfaceByAggregation<IVBEComponent> _IVBEComponent_Aggregator;
+        private IVBEComponent target_IVBEComponent { get => _IVBEComponent_Aggregator?.WrappedObject; }
+
+        private RestrictComInterfaceByAggregation<IVBETypeInfo> _IVBETypeInfo_Aggregator;
+        private IVBETypeInfo target_IVBETypeInfo { get => _IVBETypeInfo_Aggregator?.WrappedObject; }
+
+        public bool HasVBEExtensions { get => _IVBETypeInfo_Aggregator?.WrappedObject != null; }
+
+        private bool _hasModuleScopeCompilationErrors;
+        public bool HasModuleScopeCompilationErrors => _hasModuleScopeCompilationErrors;
+
+        public class FuncsCollection : IIndexedCollection<TypeInfoFunc>
+        {
+            TypeInfoWrapper _parent;
+            public FuncsCollection(TypeInfoWrapper parent) => _parent = parent;
+            override public int Count { get => _parent.Attributes.cFuncs; }
+            override public TypeInfoFunc GetItemByIndex(int index) => new TypeInfoFunc(_parent, index);
+        }
+        public FuncsCollection Funcs;
+
+        public class VarsCollection : IIndexedCollection<TypeInfoVar>
+        {
+            TypeInfoWrapper _parent;
+            public VarsCollection(TypeInfoWrapper parent) => _parent = parent;
+            override public int Count { get => _parent.Attributes.cVars; }
+            override public TypeInfoVar GetItemByIndex(int index) => new TypeInfoVar(_parent, index);
+        }
+        public VarsCollection Vars;
+
+        public class ImplementedInterfacesCollection : IIndexedCollection<TypeInfoWrapper>
+        {
+            TypeInfoWrapper _parent;
+            public ImplementedInterfacesCollection(TypeInfoWrapper parent) => _parent = parent;
+            override public int Count { get => _parent.Attributes.cImplTypes; }
+            override public TypeInfoWrapper GetItemByIndex(int index) => _parent.GetSafeImplementedTypeInfo(index);
+
+            public bool DoesImplement(string containerName, string interfaceName)
+            {
+                // check we are not runtime generated with no container
+                if (_parent.HasNoContainer()) return false;
```

Comment:
Stupid question: would "runtime generated with no container" include e.g. MSForms interfaces? IOW could this be used to get the correct interface for MSForms controls? We're currently resolving a `TextBox` as a `Control`, which is causing a number of issues.

Reply:
yes, see TypeLibsAPI.GetUserFormControlType which returns "MSForms.CommandButton" or whatever